File: jeongsiyeong/python/SWEA/treasure_password.py
T = int(input())

for test_case in range(1,T+1):
    N, K = map(int, input().split())
    str = input()
    
    L = N // 4
    
    str_extend = str + str[:L-1]
    
    nums = set()
    for i in range(N):
        h_str = str_extend[i:i+L]
        nums.add(int(h_str, 16))
    sorted_nums = sorted(list(nums), reverse=True)
    print(f"#{test_case} {sorted_nums[K-1]}")

File: jeongsiyeong/cpp/SWEA/treasure_password.cpp
#include <iostream>
#include <string>
#include <set>
#include <iterator>

using namespace std;

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);

	int T;
	cin >> T;

	for (int test_case = 1; test_case <= T; test_case++) {
		int N, K;
		cin >> N >> K;

		string str;
		cin >> str;

		int L = N / 4;

		str += str.substr(0, L - 1);

		set<int, greater<int>> nums;

		for (int i = 0; i < N; i++) {
			string h_str = str.substr(i, L);

			int num = stoi(h_str, nullptr, 16);
			nums.insert(num);
		}

		auto it = nums.begin();
		advance(it, K - 1);

		cout << "#" << test_case << " " << *it << "\n";
	}
}